Plugins/ServerPlugin.cpp

```diff
@@ -1,124 +1,285 @@
-#ifndef _WIN32_WINNT
-#define _WIN32_WINNT 0x0600  // at least windows vista required for grpc
-#endif
-
 #include "ServerPlugin.h"
-
-#include "codegen/server.grpc.pb.h"
+#include "Widgets/CodeWidget.h"
 
 #include <QDebug>
 #include <QFileDialog>
+#include <QList>
 #include <QTemporaryFile>
+#include <QTimer>
 
-#include <grpc++/channel.h>
-#include <grpc++/client_context.h>
-#include <grpc++/create_channel.h>
-#include <grpc/grpc.h>
+#include <chrono>
+#include <memory>
 
-using namespace grpc;
-using namespace buffers;
+namespace {
+
+inline std::chrono::system_clock::time_point future_deadline(size_t us) {
+  return (std::chrono::system_clock::now() + std::chrono::microseconds(us));
+}
 
-class CompilerClient {
- public:
-  explicit CompilerClient(std::shared_ptr<Channel> channel) : stub(Compiler::NewStub(channel)) {}
+void* tag(int i) { return reinterpret_cast<void*>(static_cast<intptr_t>(i)); }
+int detag(void* p) { return static_cast<int>(reinterpret_cast<intptr_t>(p)); }
 
-  void CompileBuffer(Game* game, CompileMode mode, std::string name) {
-    ClientContext context;
-    CompileRequest request;
+}  // anonymous namespace
 
-    request.mutable_game()->CopyFrom(*game);
-    request.set_name(name);
-    request.set_mode(mode);
+CallData::~CallData() { qDebug() << "~CallData"; }
 
-    std::unique_ptr<ClientReader<CompileReply> > reader(stub->CompileBuffer(&context, request));
-    CompileReply reply;
-    while (reader->Read(&reply)) {
-      qDebug() << reply.message().c_str() << reply.progress();
+template <class T>
+struct AsyncReadWorker : public CallData {
+  T element;
+  std::unique_ptr<ClientAsyncReader<T>> stream;
+
+  virtual ~AsyncReadWorker() override { qDebug() << "~AsyncReadWorker"; }
+  void operator()(const AsyncState state, const Status& /*status*/) override {
+    switch (state) {
+      case AsyncState::CONNECT: {
+        started();
+        stream->Read(&element, tag(AsyncState::READ));
+        break;
+      }
+      case AsyncState::READ: {
+        process(element);
+        stream->Read(&element, tag(AsyncState::READ));
+        break;
+      }
+      case AsyncState::FINISH: {
+        finished();
+        break;
+      }
+      default:
+        // TODO: Report error
+        break;
     }
-    Status status = reader->Finish();
   }
+  virtual void start() final { stream->StartCall(tag(AsyncState::CONNECT)); }
+  virtual void finish() final { stream->Finish(&status, tag(AsyncState::FINISH)); }
+
+  virtual void started() {}
+  virtual void finished() {}
+  virtual void process(const T& data) = 0;
+};
 
-  void CompileBuffer(Game* game, CompileMode mode) {
-    QTemporaryFile t("enigma");
-    if (!t.open()) return;
-    CompileBuffer(game, mode, t.fileName().toStdString());
+struct ResourceReader : public AsyncReadWorker<Resource> {

```

Comment:
Some of this logic is general and works for any kind of gRPC server. Some of it is specific to, eg, the compiler plugin. Why not separate the ENIGMA service code into an ENIGMA service header/source?

Reply:
What you actually mean is to separate out the generic grpc stuff into its own translation unit. It's not quite as easy as you think because, if you remember, last night I added a `LogOutput` signal to CallData and made it a `QObject` so that it can communicate stuff back through to the main window. That's sort of easy to change though, but, we don't really need to worry about this just yet because there's no real need to until the time someone wants to integrate Dejavu or something. Not really a priority for me anyway.